### PR TITLE
Implement AssemblyDependencyResolver

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Implement the `System.Runtime.Loader.AssemblyDependencyResolver` type as per dotnet/corefx#33165.

Implements dotnet/corefx#33165
Supercedes vitek-karas#8.

cc: @eerhardt @ericstj @vitek-karas 

From vitek-karas#8:

> PInvokes into hostpolicy.dll (which should live next to the runtime and thus always be reachable).
> If the PInvoke fails (missing hostpolicy.dll) we will fail for now.
> 
> Adds tests for the API into CoreCLR repo. The main reason is that with corerun
> we can easily mock the hostpolicy.dll since there's none to start with.
> Writing the same tests in CoreFX or any other place which starts the runtime through
> hostpolicy would require test-only functionality to exist in either the class itself
> or in the hostpolicy.

### CODE REVIEW

File: src/System.Private.CoreLib/System.Private.CoreLib.csproj [hunk 1]
```diff
@@ -280,6 +280,7 @@
     <Compile Include="$(BclSourcesRoot)\System\Runtime\InteropServices\UnknownWrapper.cs" />
     <Compile Include="$(BclSourcesRoot)\System\Runtime\InteropServices\VariantWrapper.cs" />
     <Compile Include="$(BclSourcesRoot)\System\Runtime\Loader\AssemblyLoadContext.cs" />
+	<Compile Include="$(BclSourcesRoot)\System\Runtime\Loader\AssemblyDependencyResolver.cs" />
```
Comment (jkotas) [inline src/System.Private.CoreLib/System.Private.CoreLib.csproj]: Nit: Indentation

File: tests/src/Loader/AssemblyDependencyResolverTests/AssemblyDependencyResolver.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
```
Comment (jkotas) [inline tests/src/Loader/AssemblyDependencyResolverTests/AssemblyDependencyResolver.cs]: License

File: src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyDependencyResolver.cs [hunk 3]
```diff
@@ -0,0 +1,309 @@
+﻿using System.Collections.Generic;
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyDependencyResolver.cs]: License

**Comment by jkoritzinsky** [comment]
@dotnet-bot test Ubuntu arm Cross Checked Innerloop Build and Test

**Comment by jkoritzinsky** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by jkoritzinsky** [comment]
I manually verified the test on my machine in multiple configurations. I'm going to merge this in.

**Comment by sandreenko** [comment]
> I manually verified the test on my machine in multiple configurations. I'm going to merge this in

Which configurations did you check?
I see that your test is failing in x86 basic configuration: [ci example](https://ci.dot.net/job/dotnet_coreclr/job/master/job/x86_checked_windows_nt/6030/).

Please exclude it until it get fixed, it makes many jobs red ([another example](https://ci.dot.net/job/dotnet_coreclr/job/master/view/x64/job/jitstress/job/x64_checked_windows_nt_r2r_gcstress15/129/testReport/)). Also, please trigger some pri1 jobs if you add a pri1 test.

**Comment by jkoritzinsky** [comment]
How do I trigger the pri1 jobs? I didn't see them on the list of optional jobs.

**Comment by sandreenko** [comment]
> How do I trigger the pri1 jobs? I didn't see them on the list of optional jobs.

All test jobs that do not have `Innerloop` in their names and are not corefx/perfomance/formatting jobs run pri1 tests.

For example: 
@dotnet-bot test Windows_NT x64 Checked Build and Test 
@dotnet-bot test Windows_NT x86 Checked Build and Test

There is a similar rule for TieredCompilation, but it is trickier.

**Pull request merged**